### ISSUE TITLE
Workflow description to get started with CloudFoundry on AWS

### ISSUE THREAD

**Issue by djvdorp** [pivotal-cf-experimental/bosh-bootloader#40]
As a follow-up on this issue: https://github.com/cloudfoundry/bosh/issues/1298#issuecomment-225778449

We would like to give this experimental new tooling a spin and see if we can get it to work, and provide you with feedback so we can get to a place where we can replace [these docs](https://docs.cloudfoundry.org/deploying/aws/index.html) with something that works sooner. 

Could you describe a workflow that should get us started with CloudFoundry on AWS?

**Comment by djvdorp** [comment]
So, we followed this general workflow to get to a point where we are stuck right now:

1. We ran `bbl unsupported-deploy-bosh-on-aws-for-concourse`, which gave us a fully functional BOSH director and AWS infrastructure
2. We created a [Cloud Foundry Deployment Manifest Stub for AWS](https://docs.cloudfoundry.org/deploying/aws/cf-stub.html) as per the instructions
3. We ensured that the `stemcell` and `cf-release v237` were present on the BOSH director
4. We ran `./scripts/generate_deployment_manifest aws cf-stub.yml > cf-deployment.yml`
5. We ran `bosh deployment cf-deployment`
6. We (tried to) run `bosh deploy`

The error we are encountering here is:

```
Deploying
---------
Are you sure you want to deploy? (type 'yes' to continue): yes

Director task 4
  Started preparing deployment > Preparing deployment. Failed: Deployment manifest should not contain cloud config properties: ["compilation", "networks", "resource_pools"] (00:00:01)

Error 190014: Deployment manifest should not contain cloud config properties: ["compilation", "networks", "resource_pools"]
```

This might be related to what @Amit-PivotalLabs said here: https://github.com/cloudfoundry/bosh/issues/1298#issuecomment-225778449

> [..] though we're waiting on one feature from BOSH that's just around the corner, to support "BOSH  1.0" style manifests, since the existing tooling for generating CF manifests only supports the 1.0 style.

Is there any way that we can move forward from here, be it a workaround or another solution, or is this just as far as we can try out this experimental new tooling?

**Comment by Amit-PivotalLabs** [comment]
Yes, that's the issue I was talking about.  The fix in BOSH is imminent, but if you don't want to wait, I've done this before:

1. Get the manifest out of the `state.json` file, and run `bbl ssh-key` to get the key.
2. Delete the director
3. Redeploy the director using the given manifest and key.

Don't apply any cloud-configs.  This will defeat part of the purpose of bbl of being able to maintain your environment, but if you're just trying to get a one-shot deployment to poke around CF, this could be okay.

Also, at some point, you may want to `bbl unsupported-create-lbs --type=cf` to create the ELBs needed by the gorouter and diego SSH proxies.  If you're going to do this, you should do this before deleting your Director and re-deploying manually.

**Comment by djvdorp** [comment]
Hi @Amit-PivotalLabs 

Thanks again for your reply -- we will try to get what you've done before to work first thing tomorrow morning when we're back in the office. We spent the last part of the day in an attempt to convert the BOSH 1.0 manifest file to a BOSH 2.0 manifest file by moving the `cloud config` properties to a separate file, but the work is pretty intensive and error-prone, and also sub-optimal obviously.


Will there be any upgrade/transformation path from BOSH 1.0 manifests to BOSH 2.0 manifests (mainly with regards to cf-release currently) ?

Or is there just going to be this fix in BOSH that will result in a BOSH version that supports both BOSH 2.0 manifests as well as (legacy) BOSH 1.0 manifests. And is there any roadmap/work-in-progress so we can see what has been done in BOSH (the fix you are referring to), in what stadium this next version (with the fix) is and/or when it would (likely) be released? We tried finding it ourselves in https://github.com/cloudfoundry/bosh, but the project is very active, so it is pretty hard for outsiders to find where this activity is happening.

**Comment by Amit-PivotalLabs** [comment]
We are working on providing officially supported BOSH-2.0-style manifests for Cloud Foundry.  And yes, part of that delivery is an upgrade path from a 1.0 manifest to a 2.0 manifest.

BOSH will support both style manifests very soon.  The work has already been accepted in the working branch of the `bosh` repo, the BOSH team simply needs to cut a final release, which they typically do every few weeks.

Another way to see what's happening with BOSH is to check the "Roadmap: Pivotal Tracker" link in the repo's README.  You can find the following story:

[enable user to use v1 and v2 manifests on the same director](https://www.pivotaltracker.com/story/show/117938521)

**Comment by djvdorp** [comment]
Hi @Amit-PivotalLabs 

We had a brief discussion today, and it makes most sense for us to wait for a little while until BOSH will support both (1.0 & 2.0) style manifests. We saw that the work has been accepted on the 20th of May, and that the (3) relevant commits are in a lot of branches, including (but not limited to) the `candidate` and `develop` branches.

Out of curiosity, could you share the git flow/git branching behavior used in the BOSH github? I assume `master` + a git `tag` is used to cut final releases, but are these taken from the `candidate` branch or from the `develop` branch? We are trying to follow the flow, hence the questions above.
We are really looking forward to having this functionality available in a new (final) BOSH version!

**Comment by Amit-PivotalLabs** [comment]
Hey @djvdorp 

Don't want to bounce you around too much, but that would be a question for the BOSH team, I don't know what their branching/promotion strategy is.  You could ask via an [issue on the `bosh` repo](https://github.com/cloudfoundry/bosh/issues), or a quick question on the [cf-bosh mailing list](https://lists.cloudfoundry.org/archives/list/cf-bosh@lists.cloudfoundry.org/) or [bosh Slack channel](https://cloudfoundry.slack.com/archives/bosh).

**Comment by djvdorp** [comment]
Hi @Amit-PivotalLabs 

I did exactly that, and will share it here for other people's reference, too:

@dpb587 told me on the bosh Slack channel:
> branching strategy = changes are made to develop. when develop passes our CI in concourse, candidate is updated to the passing commit. when we manually publish a release/stemcell, it gets committed and pushed to master and tagged; master's release commit is then merged back into develop.
> basically, all changes should be made against develop

@cppforlife told me on the bosh Slack channel:
> doc update about branching fell off my todo list. will do it today.

**Comment by djvdorp** [comment]
Hi @Amit-PivotalLabs 

Sorry for bothering you again, but did you happen to hear any update regarding this matter in terms of when this story ([enable user to use v1 and v2 manifests on the same director](https://www.pivotaltracker.com/story/show/117938521)) would be released?

The reason I am asking, is because there have been 3 final releases cut since the 14th of June (one on June 16th and two on June 17th) but the (3) relevant commits are still not in `master` and thus not released yet.
We would absolutely love to be able to wrap this up (the proper, non-workaround way) and provide you with our feedback on this journey, if possible. Again, thank you for your time and effort.

**Comment by Amit-PivotalLabs** [comment]
Hi @djvdorp 

Sorry, I don't know when that release will be available, but just like you I'm eagerly awaiting it.  I would try some of the BOSH communication channels for that specific question:

* [cf-bosh mailing list](https://lists.cloudfoundry.org/archives/list/cf-bosh@lists.cloudfoundry.org/)
* [GitHub repo issues](https://github.com/cloudfoundry/bosh/issues)
* [#bosh Slack channel](https://cloudfoundry.slack.com/messages/bosh/)

**Comment by djvdorp** [comment]
Hi @Amit-PivotalLabs 

Thanks for your reply, I got some info regarding this from @cppforlife on the bosh Slack channel:

> djvdorp [9:43 AM] ￼
> @dkalinin: I see that there have been 4 final releases cut since the 14th of June (one on June 16th, > two on June 17th and one on June 20th), but the story I am eagerly waiting to test (https://www.pivotaltracker.com/story/show/117938521) is not in a final release yet. Is there an idea on when a final release with this story in it could and would be cut?
>
> dkalinin [10:07 AM] ￼
> @djvdorp: going for this week

**Comment by djvdorp** [comment]
Hi @Amit-PivotalLabs 

Just a quick heads up -- on Jun 24, 2016, 10:54 PM GMT+2, [BOSH v257](https://github.com/cloudfoundry/bosh/releases/tag/stable-3262) got released, which includes the story we've eagerly been awaiting (https://www.pivotaltracker.com/story/show/117938521). 

We will try to proceed our work from where we left off in the coming week(s).

**Comment by djvdorp** [comment]
Hi @Amit-PivotalLabs 

Another update from our side; we have manually updated our BOSH director that was created with this tooling to BOSH v257 (3262.0), because `bosh-bootloader` is still installing a BOSH v256.2 (3232.4) director (for which I have sent a PR: https://github.com/pivotal-cf-experimental/bosh-bootloader/pull/41).

While trying to run `bosh deploy` again, I `bosh-bootloader` told me that it failed due to:
`Unknown CPI error 'Unknown' with message 'There is no ACTIVE Load Balancer named 'cfrouter''`
So we figured; maybe we need to run the command `bbl unsupported-create-lbs --type=cf` before proceeding, which we did (with a self-signed SSL certificate for testing purposes).

However, we are noticing that the ELB EC2 instances are terminated almost instantly -- could this be because we did not yet create a wildcard CNAME record in Route53 that points to the newly created ELB, and so the health checks might be failing which lead to instance termination?

PS: so far, we have written down all our feedback points, please let us know what would be an appropriate place to put them for you and the team?

**Comment by Amit-PivotalLabs** [comment]
Hi @djvdorp 

We have this story unblocked: https://www.pivotaltracker.com/story/show/120469987.  We will actually use a compiled release which will make the `bbl` flow much faster than pulling the release straight from bosh.io and having to compile.

> While trying to run `bosh deploy` again, I `bosh-bootloader` told me that it failed due to:
> `Unknown CPI error 'Unknown' with message 'There is no ACTIVE Load Balancer named 'cfrouter'`

`bosh-bootloader` is giving you that error message, or `bosh deploy`?  I suspect you only mean `bosh deploy`.  What exactly are you deploying?  I assume CF, and I assume you're using the `scripts/generate_deployment_manifest` script in `cf-release` with `aws` as your first argument.  This assumes that your ELB is called `cfrouter`.  You will need to override this in your stub to be the name `bbl` has given.  You can run `bbl unsupported-lbs` to see the ELB names.

> However, we are noticing that the ELB EC2 instances are terminated almost instantly -- could this be because we did not yet create a wildcard CNAME record in Route53 that points to the newly created ELB, and so the health checks might be failing which lead to instance termination?

What exactly is being terminated?  I assume that you're running `bosh deploy`, and then its creating several EC2 instances, including some jobs called either `router_zN` or `haproxy_zN`.  After creating the instances, BOSH then tries to attach them to the ELB called `cfrouter`, sees that there is no such ELB, and then starts rolling back by deleting those VMs.  Does this sound like an accurate description of what you're seeing?

**Comment by djvdorp** [comment]
Hi @Amit-PivotalLabs 

Thanks again for your reply! I would have loved to be able to send a compiled release as a PR too, but I am not sure how to do that, otherwise I would be more than happy to help out here!

All your assumptions above are correct, `bosh deploy` is giving me that error message, not `bosh-bootloader`. I am indeed trying to deploy CF, using the `scripts/generate_deployment_manifest` script in `cf-release` with `aws` as my first argument. I found my ELB names via `bbl lbs` and was able to override them in my `cf-stub.yml` like this:

```
resource_pools:
  - name: router_z1
    cloud_properties:
      elbs:
        - bbl-aws-2-CFRouter-PBD3OB5PYJMW

  - name: router_z2
    cloud_properties:
      elbs:
        - bbl-aws-2-CFRouter-PBD3OB5PYJMW
```

I was unsure if I had to add only the `CFRouter` ELB, or also the `CFSSHPro` ELB, but luckily you were able to answer this for me too on the bosh Slack channel and tell me that only adding the `CFRouter` is right.

Regarding the ELB EC2 instances being terminated, I am indeed running `bosh deploy` which will create several EC2 instances, `router_z1` and `router_z2` in my case. The instances were terminated because of your previous explanation regarding the missing ELB called `cfrouter`, but since I applied your above suggestion and point `bosh` to the right ELB with my name, the instances stay started (however, the ELB's do currently report them as OutOfService, but this might be the case because my `bosh deploy` is currently stuck somewhere else and has not completed successfully yet).
